### ISSUE TITLE
How to limit disk reading to only one thread to avoid unnecessary disk head trashing?

### ISSUE THREAD

**Issue by jim-collier** [sahib/rmlint#363]
I'm not sure what the correct terminology is here. By "thread" I don't necessarily mean OS or language-level concurrent thread. The main point is I only want one "thing" working on my array of four disks. The reason I'm assuming "thread" might be too specific of a term in this case, is that I could see a situation where the code alternates between, say, two code paths in rapid succession, but not multithreaded. In that hypothetical case, even though both aren't technically reading from the disk at literally the same time, the rapid alternation would produce the same kind of disk head thrashing.

So, with the semantics of the concern out of the way, here's what leads me to believe there are two operations working on my array at more-or-less the same time or close enough to the same time from the perspective of disk heads:

```
DEBUG: Starting MDS scheduler with 2 threads
DEBUG: Starting disk 2096 (pointer 0x5647ad140800) thread #1
DEBUG: Starting disk 2096 (pointer 0x5647ad140800) thread #2
```

I realize I may be misinterpreting what that means, and that two things aren't actually working on my array at more or less the same time. But on the face of it, that's the way it sounds.

And if true, then the question is: How do I limit it to only one, in order to avoid the mechanical disk heads jumping around unnecessarily, slowing things down, increasing wear, etc.?

When I run `iotop`, the rmlint's read throughput does seem much less than I would expect for sequential reads and writes, but that doesn't necessarily mean anything. Also, `iotop` does show two rmlint threads, each with roughly the same throughput:

I see that there is an option `--threads=N`. But I have some concerns with that:

- I still would presumably want multithreaded operation, e.g. to not block other things while calculating a hash.
- That wouldn't necessarily mean the disk isn't being worked on by two things at more or less the same time (e.g. rapid enough alternation to effectively be "concurrent" from a mechanical disk perspective). Though, the two threads shown in iotop kind of suggests it might.

Is there some other way to limit it to only 1 disk operation at more or less the same time, without disabling multithreaded operation altogether?

**Comment by sahib** [comment]
> Is there some other way to limit it to only 1 disk operation at more or less the same time, without disabling multithreaded operation altogether?

There will be at max one reading operation per disk at max. What you see in the output above is the hashing and the reading threads.  See also #364.

**Comment by SeeSpotRun** [comment]
There is a hidden option `--threads-per-disk`.  Set this to 1 (default is 2).
`$rmlint --threads-per-disk=1 ...`

**Comment by jim-collier** [comment]
Thank you both!

**Comment by jim-collier** [comment]
Sorry one more question. With the "--threads-per-disk=1" option, does that apply to physical disks, or logical volumes? For example, I have a 4-disk Btrfs array. I would imagine it applies to the volume, since I doubt there there's a (non-brittle) way to deal with Btrfs' inner layout. I'll test this but I'm wondering if there's a canonical answer already in your working memory. (And I do realize and appreciate that I'm asking about the workings of an undocumented feature!)

**Comment by SeeSpotRun** [comment]
Yes, it will be per volume.  If I recall / understand correctly, it will associate each volume with one physical disk (as per output of `mount`).  If there are other filesystems associated with that same physical disk, then `rmlint` won't try to read from those at the same time as from the multi-disk btrfs filesystem.  But that also means that there could be read contention between btrfs volumes and other fs's on one of the 'other' disks shared by btrfs.
TL;DR: `rmlint` tries its best but can get potentially confused by btrfs, mdadm and lvm.

**Comment by jim-collier** [comment]
Great, thanks!
